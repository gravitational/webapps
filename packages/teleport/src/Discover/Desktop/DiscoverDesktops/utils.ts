import { Path } from 'teleport/Discover/Desktop/DiscoverDesktops/lines/Path';
import { MonotoneX } from 'teleport/Discover/Desktop/DiscoverDesktops/lines/MonotoneX';

const STROKE_WIDTH = 4;

export interface Line {
  width: number;
  height: number;
  path: string;
}

export function createLine(
  desktopServiceElement: HTMLDivElement,
  desktopElement: HTMLDivElement,
  containerElement: HTMLDivElement
): Line {
  if (!desktopElement || !desktopServiceElement || !containerElement) {
    return null;
  }

  const desktopServiceRect = desktopServiceElement.getBoundingClientRect();
  const desktopRect = desktopElement.getBoundingClientRect();
  const containerRect = containerElement.getBoundingClientRect();

  const distance = desktopRect.left - desktopServiceRect.right;

  const path = new Path();
  const line = new MonotoneX(path);

  line.lineStart();

  const desktopLinePosition =
    desktopRect.top - containerRect.top + desktopRect.height / 2 - 1;
  const desktopServiceLinePosition =
    desktopServiceRect.top - containerRect.top + desktopServiceRect.height / 2;

  line.point(0, desktopServiceLinePosition - STROKE_WIDTH * 2);
  line.point(40, desktopServiceLinePosition - STROKE_WIDTH * 2);
  line.point(distance - 10, desktopLinePosition + STROKE_WIDTH / 2);
  line.point(distance, desktopLinePosition + STROKE_WIDTH / 2);

  line.lineEnd();

  return {
    width: distance,
    height: containerRect.height,
    path: path.path,
  };
}
