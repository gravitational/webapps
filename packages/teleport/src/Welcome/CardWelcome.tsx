/*
Copyright 2021 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import { Card, Box, Text, ButtonPrimary } from 'design';

export default function CardWelcome({
  title,
  subTitle,
  btnText,
  onClick,
}: Props) {
  return (
    <Card bg="primary.light" my={6} mx="auto" width="464px">
      <Box p={6}>
        <Text typography="h2" mb={3} textAlign="center" color="light">
          {title}
        </Text>
        <Text typography="h5" mb={3} textAlign="center">
          {subTitle}
        </Text>
        <ButtonPrimary width="100%" mt={3} size="large" onClick={onClick}>
          {btnText}
        </ButtonPrimary>
      </Box>
    </Card>
  );
}

type Props = {
  title: string;
  subTitle: string;
  btnText: string;
  onClick(): void;
};
