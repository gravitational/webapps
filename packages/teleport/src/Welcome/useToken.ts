/*
Copyright 2021-2022 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { useState, useEffect } from 'react';
import useAttempt from 'shared/hooks/useAttemptNext';

import cfg from 'teleport/config';
import history from 'teleport/services/history';
import auth, {
  ChangedUserAuthn,
  RecoveryCodes,
  ResetToken,
} from 'teleport/services/auth';

export default function useToken(tokenId: string) {
  const [resetToken, setResetToken] = useState<ResetToken>();
  const [recoveryCodes, setRecoveryCodes] = useState<RecoveryCodes>();
  const [success, setSuccess] = useState(false); // TODO rename
  const [privateKeyPolicyEnabled, setPrivateKeyPolicyEnabled] = useState(false);

  const fetchAttempt = useAttempt('');
  const submitAttempt = useAttempt('');
  const auth2faType = cfg.getAuth2faType();

  useEffect(() => {
    fetchAttempt.run(() =>
      auth
        .fetchPasswordToken(tokenId)
        .then(resetToken => setResetToken(resetToken))
    );
  }, []);

  function handleResponse(res: ChangedUserAuthn) {
    if (res.privateKeyPolicyEnabled) {
      setPrivateKeyPolicyEnabled(true);
    }
    if (res.recovery.createdDate) {
      setRecoveryCodes(res.recovery);
    } else {
      finishedRegister();
    }
  }

  function onSubmit(password: string, otpCode = '', deviceName = '') {
    submitAttempt.setAttempt({ status: 'processing' });
    auth
      .resetPassword({ tokenId, password, otpCode, deviceName })
      .then(handleResponse)
      .catch(submitAttempt.handleError);
  }

  function onSubmitWithWebauthn(password?: string, deviceName = '') {
    submitAttempt.setAttempt({ status: 'processing' });
    auth
      .resetPasswordWithWebauthn({ tokenId, password, deviceName })
      .then(handleResponse)
      .catch(submitAttempt.handleError);
  }

  function redirect() {
    history.push(cfg.routes.root, true);
  }

  function clearSubmitAttempt() {
    submitAttempt.setAttempt({ status: '' });
  }

  function finishedRegister() {
    setSuccess(true);
  }

  return {
    auth2faType,
    primaryAuthType: cfg.getPrimaryAuthType(),
    isPasswordlessEnabled: cfg.isPasswordlessEnabled(),
    fetchAttempt: fetchAttempt.attempt,
    submitAttempt: submitAttempt.attempt,
    clearSubmitAttempt,
    onSubmit,
    onSubmitWithWebauthn,
    resetToken,
    recoveryCodes,
    redirect,
    success,
    finishedRegister,
    privateKeyPolicyEnabled,
  };
}

export type State = ReturnType<typeof useToken>;
