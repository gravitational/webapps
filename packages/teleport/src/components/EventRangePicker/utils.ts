import moment from 'moment';

export function getRangeOptions(): EventRange[] {
  return [
    {
      name: 'Today',
      from: moment(new Date())
        .startOf('day')
        .toDate(),
      to: moment(new Date())
        .endOf('day')
        .toDate(),
    },
    {
      name: '7 days',
      from: moment()
        .subtract(6, 'day')
        .startOf('day')
        .toDate(),
      to: moment(new Date())
        .endOf('day')
        .toDate(),
    },
    {
      name: 'Custom Range...',
      isCustom: true,
      from: new Date(),
      to: new Date(),
    },
  ];
}

export type EventRange = {
  from: Date;
  to: Date;
  isCustom?: boolean;
  name?: string;
};
