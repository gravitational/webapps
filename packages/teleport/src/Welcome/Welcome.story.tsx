/*
Copyright 2021-2022 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import { MemoryRouter } from 'react-router-dom';

import Welcome from './Welcome';
import { CardWelcome } from './CardWelcome';

export default { title: 'Teleport/Welcome' };

export const WelcomeCustom = () => (
  <CardWelcome
    title="Some Title"
    subTitle="some small subtitle"
    btnText="Button Text"
    onClick={() => null}
  />
);

export const WelcomeInvite = () => (
  <MemoryRouter initialEntries={['/web/invite/1234']}>
    <Welcome />
  </MemoryRouter>
);

export const WelcomeReset = () => (
  <MemoryRouter initialEntries={['/web/reset/1234']}>
    <Welcome />
  </MemoryRouter>
);
