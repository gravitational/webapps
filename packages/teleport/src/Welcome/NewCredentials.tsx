/*
Copyright 2021 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import RecoveryCodes from 'teleport/components/RecoveryCodes';
import Form from './FormNewCredentials';
import Expired from './FormNewCredentials/Expired';
import useToken, { State } from './useToken';

export default function Container({
  tokenId = '',
  title = '',
  submitBtnText = '',
  resetMode = false,
}) {
  const state = useToken(tokenId);
  return (
    <NewCredentials
      {...state}
      title={title}
      submitBtnText={submitBtnText}
      resetMode={resetMode}
    />
  );
}

export function NewCredentials(props: Props) {
  const {
    submitAttempt,
    fetchAttempt,
    passwordToken,
    recoveryCodes,
    resetMode,
    redirect,
    ...rest
  } = props;

  if (fetchAttempt.status === 'failed') {
    return <Expired resetMode={resetMode} />;
  }

  if (fetchAttempt.status !== 'success') {
    return null;
  }

  if (recoveryCodes) {
    return (
      <RecoveryCodes
        recoveryCodes={recoveryCodes}
        redirect={redirect}
        isNewCodes={resetMode}
      />
    );
  }

  const { user, qrCode } = passwordToken;
  return <Form user={user} qr={qrCode} attempt={submitAttempt} {...rest} />;
}

export type Props = State & {
  submitBtnText: string;
  title: string;
  resetMode?: boolean;
};
