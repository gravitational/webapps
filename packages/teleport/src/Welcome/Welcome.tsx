/*
Copyright 2021 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import { Route, Switch, useParams } from 'teleport/components/Router';
import history from 'teleport/services/history';
import LogoHero from 'teleport/components/LogoHero';
import cfg from 'teleport/config';
import { NewCredentials } from './NewCredentials';
import { CardWelcome } from './CardWelcome';

export default function Welcome() {
  const { tokenId } = useParams<{ tokenId: string }>();

  const handleOnInviteContinue = () => {
    history.push(cfg.getUserInviteTokenContinueRoute(tokenId));
  };

  const handleOnResetContinue = () => {
    history.push(cfg.getUserResetTokenContinueRoute(tokenId));
  };

  return (
    <>
      <LogoHero />
      <Switch>
        <Route exact path={cfg.routes.userInvite}>
          <CardWelcome
            title="Welcome to Teleport"
            subTitle="Please click the button below to create an account"
            btnText="Get started"
            onClick={handleOnInviteContinue}
          />
        </Route>
        <Route exact path={cfg.routes.userReset}>
          <CardWelcome
            title="Reset Authentication"
            subTitle="Please click the button below to begin recovery of your account"
            btnText="Continue"
            onClick={handleOnResetContinue}
          />
        </Route>
        <Route path={cfg.routes.userInviteContinue}>
          <NewCredentials tokenId={tokenId} />
        </Route>
        <Route path={cfg.routes.userResetContinue}>
          <NewCredentials resetMode={true} tokenId={tokenId} />
        </Route>
      </Switch>
    </>
  );
}
