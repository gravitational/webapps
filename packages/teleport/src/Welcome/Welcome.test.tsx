/*
Copyright 2021-2022 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import { MemoryRouter, Route, Router } from 'react-router';
import { createMemoryHistory } from 'history';
import { screen, fireEvent, act, render, waitFor } from 'design/utils/testing';
import { Logger } from 'shared/libs/logger';
import cfg from 'teleport/config';
import history from 'teleport/services/history';
import auth from 'teleport/services/auth';
import Welcome from './Welcome';

const invitePath = '/web/invite/5182';
const inviteContinuePath = '/web/invite/5182/continue';
const resetPath = '/web/reset/5182';
const resetContinuePath = '/web/reset/5182/continue';

describe('teleport/components/Welcome', () => {
  beforeEach(() => {
    jest.spyOn(Logger.prototype, 'log').mockImplementation();
    jest.spyOn(auth, 'fetchPasswordToken').mockImplementation(async () => ({
      user: 'sam',
      tokenId: 'test123',
      qrCode: 'test12345',
    }));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should have correct welcome prompt flow for invite', async () => {
    jest.spyOn(history, 'push').mockImplementation();

    const mockHistory = createMemoryHistory({
      initialEntries: [invitePath],
    });

    render(
      <Router history={mockHistory}>
        <Route path={cfg.routes.userInvite}>
          <Welcome />
        </Route>
      </Router>
    );

    expect(
      screen.getByText(/Please click the button below to create an account/i)
    ).toBeInTheDocument();

    expect(auth.fetchPasswordToken).not.toHaveBeenCalled();

    await waitFor(() => {
      fireEvent.click(screen.getByText(/get started/i));
      mockHistory.push(inviteContinuePath);
    });

    expect(history.push).toHaveBeenCalledWith(inviteContinuePath);
    expect(auth.fetchPasswordToken).toHaveBeenCalled();

    expect(screen.getByText(/confirm password/i)).toBeInTheDocument();
  });

  it('should have correct welcome prompt flow for reset', async () => {
    jest.spyOn(history, 'push').mockImplementation();

    const mockHistory = createMemoryHistory({
      initialEntries: [resetPath],
    });

    render(
      <Router history={mockHistory}>
        <Route path={cfg.routes.userReset}>
          <Welcome />
        </Route>
      </Router>
    );

    expect(
      screen.getByText(
        /Please click the button below to begin recovery of your account/i
      )
    ).toBeInTheDocument();

    expect(auth.fetchPasswordToken).not.toHaveBeenCalled();

    await waitFor(() => {
      fireEvent.click(screen.getByText(/Continue/i));
      mockHistory.push(resetContinuePath);
    });

    expect(history.push).toHaveBeenCalledWith(resetContinuePath);
    expect(auth.fetchPasswordToken).toHaveBeenCalled();

    expect(screen.getByText(/change password/i)).toBeInTheDocument();
  });

  it('reset password', async () => {
    jest.spyOn(cfg, 'getAuth2faType').mockImplementation(() => 'off');
    jest
      .spyOn(auth, 'resetPassword')
      .mockImplementation(() => new Promise(() => null));

    await act(async () => renderInvite());

    const pwdField = screen.getByPlaceholderText('Password');
    const pwdConfirmField = screen.getByPlaceholderText('Confirm Password');

    // fill out input boxes and trigger submit
    fireEvent.change(pwdField, { target: { value: 'pwd_value' } });
    fireEvent.change(pwdConfirmField, { target: { value: 'pwd_value' } });
    fireEvent.click(screen.getByRole('button'));

    expect(auth.resetPassword).toHaveBeenCalledWith({
      tokenId: '5182',
      password: 'pwd_value',
      otpCode: '',
      deviceName: '',
    });
  });

  it('reset password with otp', async () => {
    jest.spyOn(cfg, 'getAuth2faType').mockImplementation(() => 'otp');
    jest
      .spyOn(auth, 'resetPassword')
      .mockImplementation(() => new Promise(() => null));

    await act(async () => renderInvite());

    const pwdField = screen.getByPlaceholderText('Password');
    const pwdConfirmField = screen.getByPlaceholderText('Confirm Password');
    const otpField = screen.getByPlaceholderText('123 456');

    // fill out input boxes and trigger submit
    fireEvent.change(pwdField, { target: { value: 'pwd_value' } });
    fireEvent.change(pwdConfirmField, { target: { value: 'pwd_value' } });
    fireEvent.change(otpField, { target: { value: '2222' } });
    fireEvent.click(screen.getByRole('button'));

    expect(auth.resetPassword).toHaveBeenCalledWith({
      tokenId: '5182',
      password: 'pwd_value',
      otpCode: '2222',
      deviceName: '',
    });
  });

  it('reset password with webauthn', async () => {
    jest.spyOn(cfg, 'getAuth2faType').mockImplementation(() => 'webauthn');
    jest
      .spyOn(auth, 'resetPassword')
      .mockImplementation(() => new Promise(() => null));
    jest
      .spyOn(auth, 'resetPasswordWithWebauthn')
      .mockImplementation(() => new Promise(() => null));

    await act(async () => renderInvite());

    // fill out input boxes and trigger submit
    const pwdField = screen.getByPlaceholderText('Password');
    const pwdConfirmField = screen.getByPlaceholderText('Confirm Password');
    fireEvent.change(pwdField, { target: { value: 'pwd_value' } });
    fireEvent.change(pwdConfirmField, { target: { value: 'pwd_value' } });
    fireEvent.click(screen.getByRole('button'));

    expect(auth.resetPasswordWithWebauthn).toHaveBeenCalledWith({
      tokenId: '5182',
      password: 'pwd_value',
      deviceName: '',
    });
  });

  it('reset password error', async () => {
    let reject;
    jest.spyOn(cfg, 'getAuth2faType').mockImplementation(() => 'off');
    jest.spyOn(auth, 'resetPassword').mockImplementation(() => {
      return new Promise((resolve, _reject) => {
        reject = _reject;
      });
    });

    await act(async () => renderInvite());

    // fill out input boxes and trigger submit
    const pwdField = screen.getByPlaceholderText('Password');
    const pwdConfirmField = screen.getByPlaceholderText('Confirm Password');
    fireEvent.change(pwdField, { target: { value: 'pwd_value' } });
    fireEvent.change(pwdConfirmField, { target: { value: 'pwd_value' } });

    await act(async () => {
      fireEvent.click(screen.getByRole('button'));
      reject(new Error('server_error'));
    });

    expect(screen.getByText('server_error')).toBeDefined();
  });
});

function renderInvite(url = inviteContinuePath) {
  render(
    <MemoryRouter initialEntries={[url]}>
      <Route path={cfg.routes.userInviteContinue}>
        <Welcome />
      </Route>
    </MemoryRouter>
  );
}
