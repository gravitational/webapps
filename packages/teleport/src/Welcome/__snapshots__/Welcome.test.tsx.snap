// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`teleport/components/Welcome auth type "on" should render form with hardware key as first option in dropdown 1`] = `
.c3 {
  box-sizing: border-box;
  padding: 40px;
  flex: 3;
}

.c6 {
  box-sizing: border-box;
  margin-bottom: 24px;
}

.c10 {
  box-sizing: border-box;
  max-width: 50%;
  margin-bottom: 24px;
  margin-right: 16px;
  width: 100%;
}

.c12 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  background: #512FC9;
  color: rgba(255,255,255,0.87);
  min-height: 40px;
  font-size: 12px;
  padding: 0px 40px;
  margin-top: 16px;
  width: 100%;
}

.c12:active {
  opacity: 0.56;
}

.c12:hover,
.c12:focus {
  background: #651FFF;
}

.c12:active {
  background: #354AA4;
}

.c12:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c1 {
  box-sizing: border-box;
  margin-left: auto;
  margin-right: auto;
  margin-top: 40px;
  margin-bottom: 40px;
  width: 464px;
  background-color: #222C59;
  box-shadow: 0 4px 16px rgba(0,0,0,0.24);
  border-radius: 8px;
}

.c8 {
  appearance: none;
  border: none;
  border-radius: 4px;
  box-shadow: inset 0 2px 4px rgba(0,0,0,.24);
  box-sizing: border-box;
  display: block;
  height: 40px;
  font-size: 16px;
  padding: 0 16px;
  outline: none;
  width: 100%;
  color: #324148;
  background-color: #FFFFFF;
}

.c8::-ms-clear {
  display: none;
}

.c8::placeholder {
  opacity: 0.4;
}

.c8:read-only {
  cursor: not-allowed;
}

.c7 {
  color: #FFFFFF;
  display: block;
  font-size: 11px;
  font-weight: 500;
  text-transform: uppercase;
  width: 100%;
  margin-bottom: 4px;
}

.c0 {
  display: block;
  outline: none;
  margin-left: auto;
  margin-right: auto;
  margin-top: 40px;
  margin-bottom: 40px;
  max-width: 200px;
  max-height: 120px;
}

.c4 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 300;
  font-size: 28px;
  line-height: 32px;
  margin: 0px;
  margin-bottom: 16px;
  color: #FFFFFF;
  text-align: center;
}

.c5 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 400;
  font-size: 18px;
  line-height: 32px;
  word-break: break-all;
  margin: 0px;
  margin-bottom: 16px;
}

.c2 {
  box-sizing: border-box;
  display: flex;
}

.c9 {
  box-sizing: border-box;
  display: flex;
  align-items: center;
}

.c11 .react-select-container {
  box-shadow: inset 0 2px 4px rgba(0,0,0,0.24);
  box-sizing: border-box;
  border: none;
  display: block;
  font-size: 14px;
  outline: none;
  width: 100%;
  color: rgba(0,0,0,0.87);
  background-color: #ffffff;
  margin-bottom: 0px;
  border-radius: 4px;
}

.c11 .react-select__control,
.c11 .react-select__control--is-focused {
  min-height: 40px;
  height: 40px;
  background-color: transparent;
  border-color: transparent;
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  box-shadow: none;
}

.c11 .react-select__control:hover,
.c11 .react-select__control--is-focused:hover {
  border-color: transparent;
  cursor: pointer;
}

.c11 .react-select__option:hover {
  cursor: pointer;
  background-color: #eceff1;
}

.c11 .react-select__option--is-focused {
  background-color: #eceff1;
}

.c11 .react-select__option--is-selected {
  background-color: #cfd8dc;
  color: inherit;
}

.c11 .react-select__option--is-selected:hover {
  background-color: #cfd8dc;
}

.c11 .react-select__menu {
  margin-top: 0px;
}

.c11 .react-select__indicator-separator {
  display: none;
}

.c11 .react-select__loading-indicator {
  display: none;
}

.c11 .react-select--is-disabled .react-select__single-value,
.c11 .react-select--is-disabled .react-select__placeholder {
  color: rgba(0,0,0,0.24);
}

.c11 .react-select--is-disabled .react-select__indicator {
  color: rgba(0,0,0,0.14);
}

<div>
  <img
    class="c0"
    src="file_stub"
  />
  <form
    class="c1"
    width="464px"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <div
          class="c4"
          color="light"
        >
          Welcome to Teleport
        </div>
        <div
          class="c5"
        >
          john@example.com
        </div>
        <div
          class="c6"
        >
          <label
            class="c7"
            font-size="0"
          >
            Password
          </label>
          <input
            autocomplete="off"
            class="c8"
            color="text.onLight"
            inputmode="text"
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <div
          class="c6"
        >
          <label
            class="c7"
            font-size="0"
          >
            Confirm Password
          </label>
          <input
            autocomplete="off"
            class="c8"
            color="text.onLight"
            inputmode="text"
            placeholder="Confirm Password"
            type="password"
            value=""
          />
        </div>
        <div
          class="c9"
        >
          <div
            class="c10"
            data-testid="mfa-select"
            width="100%"
          >
            <label
              class="c7"
              font-size="0"
            >
              Two-factor type
            </label>
            <div
              class="c11"
            >
              <div
                class="react-select-container css-2b097c-container"
              >
                <div
                  class="react-select__control css-yk16xz-control"
                >
                  <div
                    class="react-select__value-container react-select__value-container--has-value css-g1d714-ValueContainer"
                  >
                    <div
                      class="react-select__single-value css-1uccc91-singleValue"
                    >
                      Hardware Key
                    </div>
                    <input
                      aria-autocomplete="list"
                      class="css-62g3xt-dummyInput"
                      id="react-select-4-input"
                      readonly=""
                      tabindex="0"
                      value=""
                    />
                  </div>
                  <div
                    class="react-select__indicators css-1hb7zxy-IndicatorsContainer"
                  >
                    <span
                      class="react-select__indicator-separator css-1okebmr-indicatorSeparator"
                    />
                    <div
                      aria-hidden="true"
                      class="react-select__indicator react-select__dropdown-indicator css-tlfecz-indicatorContainer"
                    >
                      <svg
                        aria-hidden="true"
                        class="css-6q0nyr-Svg"
                        focusable="false"
                        height="20"
                        viewBox="0 0 20 20"
                        width="20"
                      >
                        <path
                          d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <button
          class="c12"
          kind="primary"
          width="100%"
        >
          Create Account
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`teleport/components/Welcome auth type "optional" should render form with hardware key as first option in dropdown 1`] = `
.c3 {
  box-sizing: border-box;
  padding: 40px;
  flex: 3;
}

.c6 {
  box-sizing: border-box;
  margin-bottom: 24px;
}

.c10 {
  box-sizing: border-box;
  max-width: 50%;
  margin-bottom: 24px;
  margin-right: 16px;
  width: 100%;
}

.c12 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  background: #512FC9;
  color: rgba(255,255,255,0.87);
  min-height: 40px;
  font-size: 12px;
  padding: 0px 40px;
  margin-top: 16px;
  width: 100%;
}

.c12:active {
  opacity: 0.56;
}

.c12:hover,
.c12:focus {
  background: #651FFF;
}

.c12:active {
  background: #354AA4;
}

.c12:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c1 {
  box-sizing: border-box;
  margin-left: auto;
  margin-right: auto;
  margin-top: 40px;
  margin-bottom: 40px;
  width: 464px;
  background-color: #222C59;
  box-shadow: 0 4px 16px rgba(0,0,0,0.24);
  border-radius: 8px;
}

.c8 {
  appearance: none;
  border: none;
  border-radius: 4px;
  box-shadow: inset 0 2px 4px rgba(0,0,0,.24);
  box-sizing: border-box;
  display: block;
  height: 40px;
  font-size: 16px;
  padding: 0 16px;
  outline: none;
  width: 100%;
  color: #324148;
  background-color: #FFFFFF;
}

.c8::-ms-clear {
  display: none;
}

.c8::placeholder {
  opacity: 0.4;
}

.c8:read-only {
  cursor: not-allowed;
}

.c7 {
  color: #FFFFFF;
  display: block;
  font-size: 11px;
  font-weight: 500;
  text-transform: uppercase;
  width: 100%;
  margin-bottom: 4px;
}

.c0 {
  display: block;
  outline: none;
  margin-left: auto;
  margin-right: auto;
  margin-top: 40px;
  margin-bottom: 40px;
  max-width: 200px;
  max-height: 120px;
}

.c4 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 300;
  font-size: 28px;
  line-height: 32px;
  margin: 0px;
  margin-bottom: 16px;
  color: #FFFFFF;
  text-align: center;
}

.c5 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 400;
  font-size: 18px;
  line-height: 32px;
  word-break: break-all;
  margin: 0px;
  margin-bottom: 16px;
}

.c2 {
  box-sizing: border-box;
  display: flex;
}

.c9 {
  box-sizing: border-box;
  display: flex;
  align-items: center;
}

.c11 .react-select-container {
  box-shadow: inset 0 2px 4px rgba(0,0,0,0.24);
  box-sizing: border-box;
  border: none;
  display: block;
  font-size: 14px;
  outline: none;
  width: 100%;
  color: rgba(0,0,0,0.87);
  background-color: #ffffff;
  margin-bottom: 0px;
  border-radius: 4px;
}

.c11 .react-select__control,
.c11 .react-select__control--is-focused {
  min-height: 40px;
  height: 40px;
  background-color: transparent;
  border-color: transparent;
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  box-shadow: none;
}

.c11 .react-select__control:hover,
.c11 .react-select__control--is-focused:hover {
  border-color: transparent;
  cursor: pointer;
}

.c11 .react-select__option:hover {
  cursor: pointer;
  background-color: #eceff1;
}

.c11 .react-select__option--is-focused {
  background-color: #eceff1;
}

.c11 .react-select__option--is-selected {
  background-color: #cfd8dc;
  color: inherit;
}

.c11 .react-select__option--is-selected:hover {
  background-color: #cfd8dc;
}

.c11 .react-select__menu {
  margin-top: 0px;
}

.c11 .react-select__indicator-separator {
  display: none;
}

.c11 .react-select__loading-indicator {
  display: none;
}

.c11 .react-select--is-disabled .react-select__single-value,
.c11 .react-select--is-disabled .react-select__placeholder {
  color: rgba(0,0,0,0.24);
}

.c11 .react-select--is-disabled .react-select__indicator {
  color: rgba(0,0,0,0.14);
}

<div>
  <img
    class="c0"
    src="file_stub"
  />
  <form
    class="c1"
    width="464px"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <div
          class="c4"
          color="light"
        >
          Welcome to Teleport
        </div>
        <div
          class="c5"
        >
          john@example.com
        </div>
        <div
          class="c6"
        >
          <label
            class="c7"
            font-size="0"
          >
            Password
          </label>
          <input
            autocomplete="off"
            class="c8"
            color="text.onLight"
            inputmode="text"
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <div
          class="c6"
        >
          <label
            class="c7"
            font-size="0"
          >
            Confirm Password
          </label>
          <input
            autocomplete="off"
            class="c8"
            color="text.onLight"
            inputmode="text"
            placeholder="Confirm Password"
            type="password"
            value=""
          />
        </div>
        <div
          class="c9"
        >
          <div
            class="c10"
            data-testid="mfa-select"
            width="100%"
          >
            <label
              class="c7"
              font-size="0"
            >
              Two-factor type
            </label>
            <div
              class="c11"
            >
              <div
                class="react-select-container css-2b097c-container"
              >
                <div
                  class="react-select__control css-yk16xz-control"
                >
                  <div
                    class="react-select__value-container react-select__value-container--has-value css-g1d714-ValueContainer"
                  >
                    <div
                      class="react-select__single-value css-1uccc91-singleValue"
                    >
                      Hardware Key
                    </div>
                    <input
                      aria-autocomplete="list"
                      class="css-62g3xt-dummyInput"
                      id="react-select-5-input"
                      readonly=""
                      tabindex="0"
                      value=""
                    />
                  </div>
                  <div
                    class="react-select__indicators css-1hb7zxy-IndicatorsContainer"
                  >
                    <span
                      class="react-select__indicator-separator css-1okebmr-indicatorSeparator"
                    />
                    <div
                      aria-hidden="true"
                      class="react-select__indicator react-select__dropdown-indicator css-tlfecz-indicatorContainer"
                    >
                      <svg
                        aria-hidden="true"
                        class="css-6q0nyr-Svg"
                        focusable="false"
                        height="20"
                        viewBox="0 0 20 20"
                        width="20"
                      >
                        <path
                          d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <button
          class="c12"
          kind="primary"
          width="100%"
        >
          Create Account
        </button>
      </div>
    </div>
  </form>
</div>
`;
